### ISSUE TITLE
NextJS AppRouter using direct URL

### ISSUE THREAD

**Issue by UbiquitousBear** [quarkiverse/quarkus-quinoa#727]
### Describe the bug

When I navigate directly to any non-root URL e.g. `http://localhost:8080/components/[id]` or `http://localhost:8080/components/`, the contents of `http://localhost:8080/` (ie `src/pages/index.tsx` is shown. If it try the same request over the nextjs dev server directly, ie `http://localhost:3000/components/[id]` the routing works as expected.

Inspecting the state of `next/router` shows that the path is picked up, but nothing else is. Unsure whether this is a next issue or a quinoa issue.

### Quinoa version

2.4.3

### Quarkus version

3.12.3

### Build / Runtime

Next.js

### Package Manager

PNPM

**Comment by melloware** [comment]
@UbiquitousBear have you read the section on SPA: https://docs.quarkiverse.io/quarkus-quinoa/dev/advanced-guides.html#spa-routing

See both the NOTE and the WARNING.

**Comment by UbiquitousBear** [comment]
@melloware 

> Note: By default, Quinoa will ignore quarkus.resteasy-reactive.path, quarkus.resteasy.path and quarkus.http.non-application-root-path path prefixes. You can specify different path prefixes to ignore using quarkus.quinoa.ignored-path-prefixes.

> Warning: Currently, for technical reasons, the Quinoa SPA routing configuration won’t work with RESTEasy Classic. Instead, you may use a workaround (if your app has all the rest resources under the same path prefix):

To clarify, I'm using the `graphql` extension, which exposes `/graphql`, so I don't think the cases fall under either, unless `/graphql` should be defined under `quarkus.quinoa.ignored-path-prefixes` ?

**Comment by melloware** [comment]
Yeah I am wondering the same thing. 

Any way you can try it out or put together a small runnable reproducer?

**Comment by melloware** [comment]
Also did you enable spa routing?   A runnable reproducer will help figure it out.

**Comment by UbiquitousBear** [comment]
@melloware I've enabled SPA routing, whilst I can't isolate the example (simply due to lack of time this week), I am more than happy to push to a private GH repo if that helps?

**Comment by melloware** [comment]
Sure it might help me if I can look at it and suggest some changes.

**Comment by UbiquitousBear** [comment]
@melloware invited :)

**Comment by melloware** [comment]
Thanks I will check it out tomorrow

**Comment by UbiquitousBear** [comment]
@melloware if you navigate directly to http://localhost:8080/post/2020/first-post/with/catch/all/routes, does it still work?

**Comment by melloware** [comment]
Ahhh it works when i click the Next links but not directly do the URL in the browser.  Let me look.

**Comment by melloware** [comment]
Interesting so when you click the link here is what Next Generates:

```
http://localhost:8080/_next/static/chunks/pages/post/%5B...slug%5D.js
```

Which is not the same as 

```
http://localhost:8080/post/2020/first-post/with/catch/all/routes
```

**Comment by melloware** [comment]
Assigned to @ia3andy i am not quite sure how this route is being lost.

**Comment by melloware** [comment]
@all-contributors add @UbiquitousBear for bug

**Comment by UbiquitousBear** [comment]
@melloware trying with 2.4.4, still not fixed?

**Comment by melloware** [comment]
No.  No fix was made I only tweaked how it discovers NextJS but not your issue which is why it's still open.

**Comment by UbiquitousBear** [comment]
Ah ok - I wonder if it's something to do with how rewriting is forwarded. I believe the same happens when I build the project into a jar for deployment.

**Comment by melloware** [comment]
Yep it's something going on with SPA routing.

**Comment by melloware** [comment]
I am debugging this. I think I am closer to seeing what is wrong.

**Comment by melloware** [comment]
Here is what I have debugged:

```
[io.qua.qui.QuinoaSPARoutingHandler] (vert.x-eventloop-thread-2) Quinoa checking SPA request '/post/2020/second-post/with/catch/all/routes'
[io.qua.qui.QuinoaSPARoutingHandler] (vert.x-eventloop-thread-2) Mount Point: '/'  Route Path: '/'
[io.qua.qui.QuinoaSPARoutingHandler] (vert.x-eventloop-thread-2) Quinoa is re-routing SPA request '/post/2020/second-post/with/catch/all/routes' to '/'
[io.qua.qui.QuinoaDevProxyHandler] (vert.x-eventloop-thread-2) Quinoa is forwarding: '/'
[io.qua.qui.QuinoaDevProxyHandler] (vert.x-eventloop-thread-2) Quinoa is forwarding: '/_next/static/chunks/webpack.js'
[io.qua.qui.QuinoaDevProxyHandler] (vert.x-eventloop-thread-3) Quinoa is forwarding: '/_next/static/development/_buildManifest.js'
[io.qua.qui.QuinoaDevProxyHandler] (vert.x-eventloop-thread-2) Quinoa is forwarding: '/_next/static/chunks/pages/_app.js'
[io.qua.qui.QuinoaDevProxyHandler] (vert.x-eventloop-thread-3) Quinoa is forwarding: '/_next/static/chunks/pages/index.js'
[io.qua.qui.QuinoaDevProxyHandler] (vert.x-eventloop-thread-2) Quinoa is forwarding: '/_next/static/development/_ssgManifest.js'
[io.qua.qui.QuinoaDevProxyHandler] (vert.x-eventloop-thread-2) Quinoa is forwarding: '/_next/static/chunks/react-refresh.js'
[io.qua.qui.QuinoaDevProxyHandler] (vert.x-eventloop-thread-2) Quinoa is forwarding: '/_next/static/chunks/main.js'
[io.net.han.cod.htt.web.WebSocketServerHandshaker] (vert.x-eventloop-thread-2) [id: 0xf37380dd, L:/127.0.0.1:8080 - R:/127.0.0.1:49346] WebSocket version V13 server handshake
[io.net.han.cod.htt.web.WebSocketServerHandshaker] (vert.x-eventloop-thread-2) WebSocket version 13 server handshake key: nsM7SH5Cyulob732tXFtWA==, response: DQ/ZZpL7GZ78wggZGGsDlPWTqoc=
[io.qua.qui.QuinoaDevWebSocketProxyHandler] (vert.x-eventloop-thread-2) Quinoa Dev WebSocket Server Connected: 127.0.0.1:3000/_next/webpack-hmr
[io.net.han.cod.htt.web.WebSocketClientHandshaker13] (vert.x-eventloop-thread-2) WebSocket version 13 client handshake key: aoCqCGY1pUICRlB533ZAkA==, expected response: r5nmJp82/Efxbzl29S9GEE1HaX4=
[io.ver.cor.htt.imp.Http1xClientConnection] (vert.x-eventloop-thread-2) WebSocket handshake complete
[io.qua.qui.QuinoaDevWebSocketProxyHandler] (vert.x-eventloop-thread-2) Quinoa Dev WebSocket Client Connected: 127.0.0.1:3000/_next/webpack-hmr
[io.qua.qui.QuinoaDevProxyHandler] (vert.x-eventloop-thread-2) Quinoa is forwarding: '/_next/static/development/_devMiddlewareManifest.json'
[io.qua.qui.QuinoaDevWebSocketProxyHandler] (vert.x-eventloop-thread-2) Quinoa Dev WebSocket Client message: {"action":"sync","hash":"172eea221fcc05a4","errors":[],"warnings":[],"versionInfo":{"staleness":"fresh","installed":"14.2.5"}}
[io.qua.qui.QuinoaDevWebSocketProxyHandler] (vert.x-eventloop-thread-2) Quinoa Dev WebSocket Server message:  {"event":"client-success","clientId":1722003263354}
[io.qua.qui.QuinoaDevWebSocketProxyHandler] (vert.x-eventloop-thread-2) Quinoa Dev WebSocket Server message:  {"event":"span-end","startTime":1722003263180,"endTime":1722003263398,"spanName":"navigation-to-hydration","attributes":{"pathname":"/post/2020/second-post/with/catch/all/routes","query":""}}
[io.qua.qui.QuinoaDevWebSocketProxyHandler] (vert.x-eventloop-thread-2) Quinoa Dev WebSocket Server message:  {"event":"ping","page":"/"}

```

This websocket message back from Next is interesting.

```
Quinoa Dev WebSocket Server message:  {"event":"span-end","startTime":1722003263180,"endTime":1722003263398,"spanName":"navigation-to-hydration","attributes":{"pathname":"/post/2020/second-post/with/catch/all/routes","query":""}}
```

**Comment by UbiquitousBear** [comment]
Line 3, why is quinoa rerouting to /?

**Comment by melloware** [comment]
@UbiquitousBear its explained right here: https://docs.quarkiverse.io/quarkus-quinoa/dev/advanced-guides.html#spa-routing

**Comment by UbiquitousBear** [comment]
Could this be a nextjs bug? But wonder why it can't be reproduced using the dev server....

**Comment by melloware** [comment]
I don't think so i think its just a matter of figuring out why Quinoa SPA handler is not handling NextJS 14 AppRouter.   I used to have it working it older NextJS versions I think 12.  But AppRouter is newer and NextJS seems to be changing every day.

**Comment by UbiquitousBear** [comment]
To clarify, the example (and where I'm see the issue) uses _Page Routing_, not _App Routing_...

**Comment by ia3andy** [comment]
I see a mention of hydratation, are you sure there is no server side rendering in there?

**Comment by UbiquitousBear** [comment]
> I see a mention of hydratation, are you sure there is no server side rendering in there?

what are you defining as hydration? Hydration is happening at some level, always does in React applications.

**Comment by ia3andy** [comment]
@UbiquitousBear AFAIK hydration is happening only when the server is doing part of the work which is not the case for plain SPA no?

**Comment by UbiquitousBear** [comment]
> @UbiquitousBear AFAIK hydration is happening only when the server is doing part of the work which is not the case for plain SPA no?

That's one definition of hydration when applied to the App Router. Hydration as-is, in client side generated apps still refers to attaching event listeners to the static HTML, making it dynamic at some level.

**Comment by UbiquitousBear** [comment]
@melloware ant luck with this, it’s slowly becoming a blocker, but I can’t also be the only one with this issue?

**Comment by melloware** [comment]
see this discussion: https://github.com/quarkiverse/quarkus-quinoa/discussions/788

@zZHorizonZz looked into this and he is running into the same conclusions we are.

**Comment by zZHorizonZz** [comment]
Just wanted to provide an update, if anyone is interested: there seems to be some work being done on this issue in the [Next.js repository](https://github.com/vercel/next.js/discussions/55393#discussioncomment-10262206). In terms of routing, this might be supported by Next.js in the future, but there could still be other issues, or it could be fixed entirely, we'll need to wait and see what they come up with.

**Comment by melloware** [comment]
thanks for @LSRaGeUx release 2.8.1 with this fix and a new property read all about it: https://github.com/quarkiverse/quarkus-quinoa/pull/1015
